federation/executor.go

```diff
@@ -39,65 +43,41 @@ func fetchSchema(ctx context.Context, e ExecutorClient, optionalArgs interface{}
 }
 
 func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Executor, error) {
-	return NewExecutorWithOptionalArgs(ctx, executors, nil)
+	return NewCustomExecutor(ctx, executors, nil, nil, 5)
 }
 
-func NewExecutorWithOptionalArgs(ctx context.Context, executors map[string]ExecutorClient, optionalArgs interface{}) (*Executor, error) {
-	// Fetches the schemas from the executors clients
-	schemas := make(map[string]*introspectionQueryResult)
-	for server, client := range executors {
-		schema, err := fetchSchema(ctx, client, optionalArgs)
-		if err != nil {
-			return nil, oops.Wrapf(err, "fetching schema %s", server)
-		}
-
-		var iq introspectionQueryResult
-		if err := json.Unmarshal(schema, &iq); err != nil {
-			return nil, oops.Wrapf(err, "unmarshaling schema %s", server)
-		}
-
-		schemas[server] = &iq
+func NewCustomExecutor(ctx context.Context, executors map[string]ExecutorClient, schemaSyncer SchemaSyncer, optionalArgs interface{}, schemaSyncInterval int64) (*Executor, error) {
+	if schemaSyncer == nil {
+		schemaSyncer = newIntrospectionSchemaSyncer(ctx, executors, optionalArgs)
 	}
-
-	types, err := convertSchema(schemas)
-	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
+	if schemaSyncInterval < 1 {
+		schemaSyncInterval = 1
 	}
-
-	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
-	introspectionServer := &Server{schema: introspectionSchema}
-
-	executors["introspection"] = &DirectExecutorClient{Client: introspectionServer}
-	schema, err := introspection.RunIntrospectionQuery(introspection.BareIntrospectionSchema(introspectionServer.schema))
-
-	var iq introspectionQueryResult
-	if err := json.Unmarshal(schema, &iq); err != nil {
-		return nil, oops.Wrapf(err, "unmarshaling introspection schema")
-	}
-
-	schemas["introspection"] = &iq
-	types, err = convertSchema(schemas)
+	_, planner, err := schemaSyncer.UpdatePlanner(ctx, nil)
 	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
+		return nil, oops.Wrapf(err, "failed to load schema")
 	}
-
-	flattener, err := newFlattener(types.Schema)
-	if err != nil {
-		return nil, oops.Wrapf(err, "flattening schemas error")
+	executor := &Executor{
+		Executors:    executors,
+		planner:      planner,
+		ticker:       time.NewTicker(time.Duration(schemaSyncInterval) * time.Second),
+		schemaSyncer: schemaSyncer,
 	}
+	go executor.poll(ctx)
+	return executor, nil
+}
 
-	// The planner is aware of the merged schema and what executors
-	// know about what fields
-	planner := &Planner{
-		schema:    types,
-		flattener: flattener,
+func (e *Executor) poll(ctx context.Context) {
+	for {
+		select {
+		case <-e.ticker.C:
+			schema, newPlanner, err := e.schemaSyncer.UpdatePlanner(ctx, e.schema)
+			if err == nil && schema != nil && newPlanner != nil {
+				e.schema = schema
+				e.planner = newPlanner
```

Comment:
what will happen to the inflight queries?

Reply:
should we add a lock around the planner?

Reply:
discussed offline - adding a mutex lock around the planner